### PR TITLE
ShiftRegister() and ShiftRegisters() now has a flatten option

### PR DESCRIPTION
**Pull request by oharboe**
flattening is default on, when off a module is used to implement
the shift registers.

### Contributor Checklist

- [ ] Did you add Scaladoc to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [ ] Did you delete any extraneous printlns/debugging code?
- [ ] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [ ] Did you state the API impact?
- [ ] Did you specify the code generation impact?
- [ ] Did you request a desired merge strategy?
- [ ] Did you add text to be included in the Release Notes for this change?

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (Bug fix: `3.3.x`, [small] API extension: `3.4.x`, API modification or big change: `3.5.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you mark as `Please Merge`?

### CODE REVIEW

**Comment by sequencer** [comment]
I don't object adding `Module` to `ShiftRegister(s)`, this is good for post synthesis flow, since PnR may need take additional care to this.
I personally think there are two method to do so:
1. adding a module and provide a `inline` option, but default to be `false` to avoid affecting other users. add `firrtl.passes.InlineAnnotation` to the correspond instance if true.
2. rather than adding a new `Module`, you can also use `chisel3.util.experimental.group` to gather register into a Module.

File: src/main/scala/chisel3/util/Reg.scala [hunk 1]
```diff
@@ -79,6 +91,20 @@ object ShiftRegisters
...
     */
-  def apply[T <: Data](in: T, n: Int, resetData: T, en: Bool): Seq[T] =
-    Seq.iterate(in, n + 1)(util.RegEnable(_, resetData, en)).drop(1)
+  def apply[T <: Data](in: T, n: Int, resetData: T, en: Bool, flatten: Boolean): Seq[T] = {
+    if (flatten) {
+      Seq.iterate(in, n + 1)(util.RegEnable(_, resetData, en)).drop(1)
+    } else  {
+      class ShiftRegistersX extends Module {
+        val resetData = IO(Input(Bool()))
+        val in = IO(Input(UInt()))
```
Comment (mwachs5) [inline src/main/scala/chisel3/util/Reg.scala:100]: This should not be an IO

File: src/main/scala/chisel3/util/Reg.scala [hunk 2]
```diff
@@ -68,8 +68,20 @@ object ShiftRegisters
     * @param en enable the shift
     *
     */
-  def apply[T <: Data](in: T, n: Int, en: Bool = true.B): Seq[T] =
-    Seq.iterate(in, n + 1)(util.RegEnable(_, en)).drop(1)
+  def apply[T <: Data](in: T, n: Int, en: Bool = true.B, flatten: Boolean = true): Seq[T] = {
+    if (flatten) {
+      Seq.iterate(in, n + 1)(util.RegEnable(_, en)).drop(1)
+    } else  {
+      class ShiftRegisters extends Module {
```
Comment (mwachs5) [inline src/main/scala/chisel3/util/Reg.scala:75]: Can you call this ShiftRegister or ShiftReg? The "s" is implied

File: src/main/scala/chisel3/util/Reg.scala [hunk 3]
```diff
@@ -42,7 +42,7 @@ object ShiftRegister
     * val regDelayTwo = ShiftRegister(nextVal, 2, ena)
     * }}}
     */
-  def apply[T <: Data](in: T, n: Int, en: Bool = true.B): T = ShiftRegisters(in, n, en).lastOption.getOrElse(in)
+  def apply[T <: Data](in: T, n: Int, en: Bool = true.B, flatten: Boolean = true): T = ShiftRegisters(in, n, en, flatten).lastOption.getOrElse(in)
```
Comment (mwachs5) [inline src/main/scala/chisel3/util/Reg.scala:45]: I think "inline" is a better name for this parameter, it's more industry standard

File: src/main/scala/chisel3/util/Reg.scala [hunk 4]
```diff
@@ -79,6 +91,20 @@ object ShiftRegisters
...
     *
     */
-  def apply[T <: Data](in: T, n: Int, resetData: T, en: Bool): Seq[T] =
-    Seq.iterate(in, n + 1)(util.RegEnable(_, resetData, en)).drop(1)
+  def apply[T <: Data](in: T, n: Int, resetData: T, en: Bool, flatten: Boolean): Seq[T] = {
+    if (flatten) {
+      Seq.iterate(in, n + 1)(util.RegEnable(_, resetData, en)).drop(1)
+    } else  {
+      class ShiftRegistersX extends Module {
+        val resetData = IO(Input(Bool()))
```
Comment (mwachs5) [inline src/main/scala/chisel3/util/Reg.scala:99]: Doesn't this need en as an input too?

File: src/main/scala/chisel3/util/Reg.scala [hunk 5]
```diff
@@ -68,8 +68,20 @@ object ShiftRegisters
...
     *
     */
-  def apply[T <: Data](in: T, n: Int, en: Bool = true.B): Seq[T] =
-    Seq.iterate(in, n + 1)(util.RegEnable(_, en)).drop(1)
+  def apply[T <: Data](in: T, n: Int, en: Bool = true.B, flatten: Boolean = true): Seq[T] = {
+    if (flatten) {
+      Seq.iterate(in, n + 1)(util.RegEnable(_, en)).drop(1)
+    } else  {
+      class ShiftRegisters extends Module {
+        val in = IO(Input(UInt()))
```
Comment (mwachs5) [inline src/main/scala/chisel3/util/Reg.scala:76]: Doesn't this need en as an input too?

**Review by mwachs5 — CHANGES_REQUESTED** [review]
You will need to write tests

**Comment by mwachs5** [comment]
I also like the @sequencer suggestion of the inlineAnnotation, but we should look at the difference in the output from the current solution to one using that to see if the changes are acceptable

**Comment by oharboe** [comment]
> You will need to write tests

Yep. This is just for discussion for now though.

I want to know how to handle indexing of vectors, where I think it would be similarly helpful to control flattening.

**Comment by oharboe** [comment]
This needs a more cross-cutting solution. In the spirit of this pull request, there should be a way to control flattening of  Vec() index access. Vec() access for something like Vec(16, UInt(64.W)) generates a giant fog of logic and should be representable as a module.

**Pull request closed without merging**